lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+                                dep_task.start, dep_task.attempts,
+                                dep_task.status)
+            else:
+                tty.debug('{0} has no build task to update for {1}\'s success'
+                          .format(dep_id, pkg_id))
+
+    def install(self, **kwargs):
+        """
+        Install the package and or associated dependencies.
+
+        Args:"""
+
+        install_deps = kwargs.get('install_deps', True)
+        keep_prefix = kwargs.get('keep_prefix', False)
+        keep_stage = kwargs.get('keep_stage', False)
+        restage = kwargs.get('restage', False)
+
+        # install_self defaults True and is popped so that dependencies are
+        # always installed regardless of whether the root was installed
+        install_self = kwargs.pop('install_package', True)
+
+        # TODO: Should this be first or after package updates?
+        # TODO: It was being done AFTER external, upstream, and already
+        # TODO:   installed checks in _do_install_pop_kwargs() (i.e., just
+        # TODO:   before recursively installing dependencies and or bootstrap
+        # TODO:   compiler.
+        self._check_last_phase(**kwargs)
+
+        # Skip out early if the spec is not being installed locally (i.e., if
+        # external or upstream).
+        if not _check_install_locally(self.pkg, True):
+            return
+
+        # Initialize the build task queue
+        self._init_queue(install_deps, install_self)
+
+        # Proceed with the installation
+        while self.build_pq:
+            task = self._pop_task()
+            if task is None:
+                continue
+
+            pkg, spec = task.pkg, task.pkg.spec
+            pkg_id = pkg.unique_id
+            tty.verbose('Processing {0}: task={1}'.format(pkg_id, task))
+
+            # Ensure that the current spec has NO uninstalled dependencies,
+            # which is assumed to be reflected directly in its priority.
+            #
+            # If the spec has uninstalled dependencies, then there must be
+            # a bug in the code (e.g., priority queue or uninstalled
+            # dependencies handling).  So terminate under the assumption that
+            # all subsequent tasks will have non-zero priorities or may be
+            # dependencies of this task.
+            if task.priority != 0:
+                tty.error('Detected uninstalled dependencies for {0}: {1}'
+                          .format(pkg_id, task.uninstalled_deps))
+                dep_str = 'dependencies' if task.priority > 1 else 'dependency'
+                raise InstallError(
+                    'Cannot proceed with {0}: {1} uninstalled {2}: {3}'
+                    .format(pkg_id, task.priority, dep_str,
+                            ','.join(task.uninstalled_deps)))
+
+            # TODO: Add a check to ensure no attempt is made to install the pkg
+            # TODO:   before any of its dependencies?
```

Comment:
Highlighting TODO

Reply:
This was something Todd suggested at one point.  This should not be able to happen with the priority queue (at least not as currently implemented).